### ISSUE TITLE
ADAM updates

### ISSUE THREAD

**Issue by skaae** [Lasagne/Lasagne#144]
Implementation of "Adam: A Method for Stochastic Optimization"  http://arxiv.org/abs/1412.6980

Its adapted from https://gist.github.com/Newmu/acb738767acb4788bac3

How do I handle the license/copyright? 

I added `gamma` which seemed to be missing?

How do i initialize i as a scalar without `np.float32`?

I'll create PR if there is interest?

```PYTHON
def adam(loss, all_params, learning_rate=0.0002, beta1=0.1, beta2=0.001, 
         epsilon=1e-8, gamma=1-1e-8):
    updates = []
    all_grads = theano.grad(loss,all_params)

    i = theano.shared(np.float32(1))  # HOW to init scalar shared?
    i_t = i + 1.
    fix1 = 1. - (1. - beta1)**i_t
    fix2 = 1. - (1. - beta2)**i_t
    beta1_t = 1-(1-beta1)*gamma**(i_t-1)   # ADDED
    learning_rate_t = learning_rate * (T.sqrt(fix2) / fix1)

    for param_i, g in zip(all_params, all_grads):
        m = theano.shared(
            np.zeros(param_i.get_value().shape, dtype=theano.config.floatX))
        v = theano.shared(
            np.zeros(param_i.get_value().shape, dtype=theano.config.floatX))

        m_t = (beta1_t * g) + ((1. - beta1_t) * m) # CHANGED from b_t to use beta1_t
        v_t = (beta2 * g**2) + ((1. - beta2) * v)
        g_t = m_t / (T.sqrt(v_t) + epsilon)
        param_i_t = param_i - (learning_rate_t * g_t)

        updates.append((m, m_t))
        updates.append((v, v_t))
        updates.append((param_i, param_i_t) )
    updates.append((i, i_t))
    return updates

```

EDIT: Update parameters to match Sanders suggestions

**Comment by ebenolson** [comment]
I've also been playing with this (https://github.com/ebenolson/Lasagne/commit/2be3e7d70a408a4af18aab4e2c04e0c23d6e2594), but I haven't seen much advantage over rmsprop yet. Do you understand the reason for gamma? It seems like it vanishes immediately with the default setting, and I haven't noticed any effect from adding it. 

I initialized scalars like this, but I don't know if it's the right way:
`theano.shared(np.dtype(theano.config.floatX).type(1))`

**Comment by skaae** [comment]
Gamma is an addition to version 2 of the paper on arxiv.
I think gamma is just shrinking Beta1 over timer which corresponds to
increasing the momentum?

2015-02-19 20:47 GMT+01:00 Eben Olson <notifications@github.com>:

**Comment by benanne** [comment]
What you call gamma is actually called lambda in the paper. Also there's a mistake in the paper, and the initial value should actually be 1 - 1e-8 instead of 1e-8. Otherwise the effective beta1 jumps to 1.0 immediately.

That said, with lambda = 1 - 1e-8, it takes ages for beta1 to actually change (on the order of 1e7 update steps). So a value with a smaller exponent is probably better for most problems.

I've been talking to one of the authors about this and he said a 3rd version of the paper is coming soon, so hopefully that will clear up a few things.

**Comment by skaae** [comment]
Yes lambda, I messed up with the greek letters.

Good point with the initial value.

2015-02-19 23:12 GMT+01:00 Sander Dieleman <notifications@github.com>:

**Comment by skaae** [comment]
It's fine to stick with your version.  
So setting lambda to someting less than one corresponds to increasing beta1 after each epoch? 

Do you know what the interpretation of that is?

**Comment by benanne** [comment]
Not after each epoch, but rather after each update (by a very small amount). The idea is to slowly reduce momentum, which is supposed to improve convergence at the end of training. I talked to some people on Twitter about it earlier today: https://twitter.com/ogrisel/status/568316753593434112

**Comment by skaae** [comment]
Thanks for the link. After each update makes sense. Don't you usually increase the momentum and lower the learning rate at the later stages of training?

**Comment by benanne** [comment]
I don't tend to do that, but yeah, that's what I've seen people do before. So this is all a little counterintuitive. Hopefully the third version of the paper will clear up some of the confusion.

**Comment by skaae** [comment]
Ok, that's at least what Hinton has been preaching :) thanks for the help

**Comment by benanne** [comment]
To clarify, I do tend to decrease the learning rate, but I usually don't touch momentum. Maybe I should :)

**Comment by skaae** [comment]
I just tested it at it seems to give me way better results than adadelta

2015-02-20 0:04 GMT+01:00 Sander Dieleman <notifications@github.com>:

**Comment by benanne** [comment]
What did you test exactly? adam v2? Increasing momentum? Decreasing momentum?

**Comment by skaae** [comment]
I just switched from adadelta with Lasagne standard settings to the ADAM code with your suggestions. 

Settings:

1.  learning_rate=0.0002
2.  beta1=0.1
3.  beta2=0.001
4.  epsilon=1e-8
5.  gamma=1-1e-7

For biological sequence analysis using LSTM it seemed to give me a few percent points reduction in the error.

**Comment by gaoyuankidult** [comment]
@skaae Have you made PR request yet ? It is very useful algorithm

**Comment by skaae** [comment]
Nope, I think we agreed to wait until v3 of the paper?

I put the code on gist if you want to use it: https://gist.github.com/skaae/2fcd1c6d6d30f5fe9243

best regards Søren

**Comment by KarnUllrich** [comment]
In case it adds to the discussion, I have implemented ADAM from the paper V2, too.

It considers the mistake from the paper regarding lambda and is in line with the paper notation and the other update methods in Lasagne.

https://github.com/benanne/Lasagne/compare/master...KarnUllrich:ADAM_Op

If wished I can created a pull request.

Cheers,
Karen

**Comment by benanne** [comment]
cool! I'd still like to wait for v3 of the paper before we merge anything though... even with the 'fix' I'm not entirely convinced that the algorithm works as intended.

**Comment by craffel** [comment]
They posted v4 of the paper today:
http://arxiv.org/pdf/1412.6980v4.pdf

**Comment by skaae** [comment]
I created a gist. 
https://gist.github.com/skaae/2aa50af4efb20a745ad4

I can you check it or I can create a PR first?
I'll remove the long comments in the PR

**Comment by skaae** [comment]
Updated code:

https://gist.github.com/skaae/ae7225263ca8806868cb

**Comment by KarnUllrich** [comment]
Thanx craffel! 

Did you recognize there is again some typo/error in it? Or what is $ beta_2^t$ supposed to mean?

(see how to Compute bias-corrected second raw moment estimate)

**Comment by benanne** [comment]
After reading the new version of the paper, I have to say I'm disappointed. There is still almost nothing in the text about the parameter lambda and its purpose.  Why is this needed? The first version of the paper didn't have it and the algorithm seemed to work fine.

Looking at the algorithm description on page 2, it's supposed to "decay the first moment running average coefficient" beta1. But if the suggested value for lambda is used (1-1e8), no meaningful decay occurs. It would require about 10**7 update steps to see any meaningful decay of beta1. In all of their experiments on MNIST, the number of performed update steps is orders of magnitude smaller (about 400 updates per epoch, and they do about 200 epochs at most, so 80000 updates at most).

So in summary this additional hyperparameter lambda seems completely pointless to me. I feel like I must be missing something.

**Comment by craffel** [comment]
That's a good point.  Without lambda (or with the suggested value of lambda, for that matter), it's pretty similar to Graves' RMSProp http://arxiv.org/pdf/1308.0850v5.pdf (eqn 38-41) except that Graves doesn't include the bias correction terms and applies the first-order moment estimate differently (in the denominator, subtracted from the second order estimate within the square root instead of in the numerator).  I get why the bias correction is important, but I don't have any intuition as to why they apply the first order moment differently.

**Comment by benanne** [comment]
After talking to the authors it became clear that the lambda parameter was introduced to be able to prove convergence. In practice it does almost nothing and is not really needed. I still think it would have been useful to mention that in the paper itself :)

Adam is being used a lot so it'd be good to have a version of it in the library. I think many people have their own implementations by now, if anyone wants to send a PR, go for it.

**Comment by skaae** [comment]
I have one and can create af PR.

**Comment by skaae** [comment]
You can probably close this now?

**Comment by benanne** [comment]
Yep!
